### PR TITLE
async logic part 1 - dryrun logic

### LINKED ISSUE #101: async logic
atm, the bot logic runs through orders one by one, or one by one bundled based on pair, in a sync manner, this is not really scalable and desirable, as number of orders increase a single loop through them would take time and orders mayb lose opps once they are checked again on next run, and one can specify more issues with such design.

a possible solution is to make the logic async, ie check orders in an async manner, their vault balances (using multicall), getting eth price, etc. and just filter out the non opps. and then execute each found app one by one, and if the time between 2 opps clearing is significant, the simulation logic can be executed for the next opps again to make sure they are still valid.

this means that at times there will be burst rpc calls, and 3rd party providers might ratelimit or not allow infinite number of concurrent calls, so there should some sort of logic that number of concurrent calls can be managed reasonably within the acceptable range of the rpc provider.

a solution for this is to process orders based on token pair in an async manner, that would be:
- sort all orders based on token pair, then loop over each pair (like all order with input USDT and output DAI)
- get all vault balances for all orders in a token pair with a multicall
- exclude those with 0 balance
- get pool details for that pair (this will be single fetch compared to doing it individually for each order in previous logic)
- get gas price (this will be single fetch compared to doing it individually for each order in previous logic)
- get eth price (this will be single fetch compared to doing it individually for each order in previous logic)
- now remaining order can be tried for an `arb()` async, and only those that succeed will proceed to be submitted as a tx onchain, something to note here, is if a single submission take more than reasonable time, then the next opps should be reevaluated, another thing is to make sure running orders asyn wont surpass the concurrency limits of the rpc provider
- repeat above steps for orders of next token pair

this will be a relatively easy and small change, and can be achieved in a short time, but its impact on efficiency and performance will be noticeable

todo:
- [x] - #115 submitted
- [ ] - #115 merged
- [x] - #116 submitted
- [ ] - #116 merged
- [x] - #117 submitted
- [ ] - #117 merged
- [x] - #118 submitted
- [ ] - #118 merged
- [x] - #119 submitted
- [ ] - #119 merged

### PR DESCRIPTION
**Pull request by rouzwelt**
## Motivation
part 1 of implementing async logic.
resolves #101 

## Solution
setup dryrun logic into its own file with the required refactors.
- the new `dryrun()` fn maps to previous `dryrun()`:
https://github.com/rainlanguage/arb-bot/blob/7915cf1b92c4645209a2c980ca5c9eb3f425a13c/src/processOrders.js#L604
with almost no new changes but now with its own unit tests.

- the new `dryrunWithRetries()` fn maps to previous logic starting from:
https://github.com/rainlanguage/arb-bot/blob/7915cf1b92c4645209a2c980ca5c9eb3f425a13c/src/processOrders.js#L373
to: 
https://github.com/rainlanguage/arb-bot/blob/7915cf1b92c4645209a2c980ca5c9eb3f425a13c/src/processOrders.js#L434
but now as a sep fn and with its own unit tests, but with some minor changes on top.

## Checks

By submitting this for review, I'm confirming I've done the following:
- [x] made this PR as small as possible
- [x] unit-tested any new functionality
- [x] linked any relevant issues or PRs

### CODE REVIEW

File: src/processes/dryrun.js [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+const DryrunHaltReason = {
+    NoOpportunity: 1,
+    NoWalletFund: 2,
+    NoRoute: 3,
+};
+
+/**
+ * Tries to find the maxInput for an arb tx by doing a binary search
+ */
+async function dryrun({
```
Comment (thedavidmeister) [inline src/processes/dryrun.js:17]: does this replace something that already exists? i would expect to see some deletions that map to these additions

Comment (rouzwelt) [reply src/processes/dryrun.js:17]: this specific fn han no changes, but you can check the diff against src/processOrders.js dryrun fn

Comment (rouzwelt) [reply src/processes/dryrun.js:17]: also if I do the deletions, bunch of things will be broken, is that ok with you?

Comment (rouzwelt) [reply src/processes/dryrun.js:17]: also `dryrunWithRetries()` is a new fn which is refactor based on certain part of src/processOrders.js:processPair()

File: src/processes/dryrun.js [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+                    hopAttrs["error"] = getSpanException(error);
+                    // reason, code, method, transaction, error, stack, message
+                }
+            }
+        }
+        // record this hop attributes
+        allHopsAttributes.push(JSON.stringify(hopAttrs));
+
+        // set the maxInput for next hop
+        maximumInput = binarySearchLastHopSuccess
```
Comment (thedavidmeister) [inline src/processes/dryrun.js:259]: it should be possible to abstract out binary search logic

**Review by thedavidmeister — CHANGES_REQUESTED** [review]

File: src/processes/dryrun.js [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+        return result;
+    }
+}
+
+/**
+ * Tries to find an opp by doing a binary search for the maxInput of an arb tx
+ * it calls dryrun() on each iteration and based on the outcome, +/- the maxInput
+ * until the binary search is over and returns teh final result
+ */
+async function findOpp({
```
Comment (thedavidmeister) [inline src/processes/dryrun.js:229]: can we use the binary search logic here for the existing binary search?

**Review by thedavidmeister — CHANGES_REQUESTED** [review]

**Pull request closed without merging**